lib/time/counter.c

```diff
@@ -0,0 +1,305 @@
+		return ret;
+	}
+
+	counter_center_bottom_ticks = top_ticks >> 2;
+	counter_center_top_ticks = counter_center_bottom_ticks * 3;
+
+	if (counter_ticks < counter_center_bottom_ticks) {
+		get_zero_ticks(ticks);
+		*ticks += counter_ticks;
+	} else if (counter_ticks > counter_center_top_ticks) {
+		get_zero_ticks(ticks);
+		*ticks -= top_ticks;
+		*ticks += counter_ticks;
+	} else {
+		get_center_ticks(ticks);
+		*ticks = center_ticks;
+		*ticks += counter_ticks;
+	}
+
+	return 0;
+}
+
+static void semlock_lock(void)
+{
+	(void)k_sem_take(&semlock, K_FOREVER);
+}
+
+static void semlock_unlock(void)
+{
+	k_sem_give(&semlock);
+}
+
+static void alarm_callback(const struct device *dev,
+			   uint8_t chan_id,
+			   uint32_t ticks,
+			   void *user_data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(chan_id);
+	ARG_UNUSED(ticks);
+	ARG_UNUSED(user_data);
+
+	advance_zero_ticks();
+}
+
+static int set_center_alarm(void)
+{
+	const struct counter_alarm_cfg config = {
+		.callback = alarm_callback,
+		.ticks = top_ticks >> 1,
+		.user_data = NULL,
+		.flags = COUNTER_ALARM_CFG_ABSOLUTE,
+	};
+
+	return counter_set_channel_alarm(counter, CENTER_ALARM_ID, &config);
+}
+
+static void cancel_center_alarm(void)
+{
+	(void)counter_cancel_channel_alarm(counter, CENTER_ALARM_ID);
+}
+
+static void top_callback(const struct device *dev, void *user_data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(user_data);
+
+	advance_center_ticks();
+	(void)set_center_alarm();
+}
+
+static int set_counter_top(void)
+{
+	const struct counter_top_cfg config = {
+		.ticks = top_ticks,
+		.callback = top_callback,
+		.user_data = NULL,
+		.flags = 0,
+	};
+
+	return counter_set_top_value(counter, &config);
+}
+
+static int reconfigure_counter(void)
+{
+	int ret;
+
+	cancel_center_alarm();
+
+	ret = set_counter_top();
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = set_center_alarm();
+	if (ret < 0) {
+		return ret;
+	}
+
+	reset_ticks();
+	return ret;
+}
+
+static int configure_counter(void)
+{
+	int ret;
+
+	ret = device_is_ready(counter) ? 0 : -ENODEV;
+	if (ret < 0) {
+		return ret;
+	}
+
+	frequency = counter_get_frequency(counter);
+	top_ticks = counter_get_max_top_value(counter);
+
+	ret = set_counter_top();
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = set_center_alarm();
+	if (ret < 0) {
+		return ret;
+	}
+
+	reset_ticks();
+
+	ret = counter_start(counter);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static void add_ticks_to_seconds(int64_t *seconds, const uint64_t *ticks)
+{
+	*seconds += (int64_t)((*ticks) / frequency);
+}
+
+int sys_time_get_timestamp(int64_t *timestamp)
+{
+	int ret;
+	uint64_t ticks;
+
+	ret = reference_is_set() ? 0 : -ENODATA;
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = get_ticks(&ticks);
+	if (ret < 0) {
+		return ret;
+	}
+
+	semlock_lock();
+	get_reference(timestamp);
+	semlock_unlock();
+	add_ticks_to_seconds(timestamp, &ticks);
+	return sys_time_validate_timestamp(timestamp) ? 0 : -EAGAIN;
+}
+
+int sys_time_set_timestamp(const int64_t *timestamp)
```

Comment:
The specific kind of counter and conversion assumptions should be decoupled from sys_time. Either make it clear where future decoupling can be done (even if not required now) or name the function more specifically and place it somewhere close to the specific type of counter API you're expecting here.

Reply:
I don't agree with this comment. The counter API has no relevance to the sys_time API, neither does the struct rtc_time have any relevance to the RTC API. Implementations of the sys_time API may use either, but the entire point is to abstract this away from the user, who simply wants to get/set and convert time.

I understand that the fact that we are reusing the `struct rtc_time` makes it seem like this is directly related to the rtc device driver API, it is not, `sys_time` works just fine on systems that don't have an RTC...